### ISSUE TITLE
`no-unresolved` JS alias regression in 3.8+

### ISSUE THREAD

**Issue by RobHannay** [import-js/eslint-import-resolver-typescript#363]
Since 3.8 aliases are no-longer being resolved in `.js` files.

I've read a few of the issues raised about 3.8, but I think this is separate. I have tested with 3.8.3 and still face the issue.

We use aliases that are actually defined in three possible places:
- webpack.config
- jsconfig
- tsconfig

![Image](https://github.com/user-attachments/assets/e83ed009-0377-40ec-af69-b33f1a026209)

Perhaps it was never intended to resolve imports _from_ js files?

It's fine is I use relative paths with lots of `../`.

We have a verbose config but the key details are:

```ts
{
  eslint.configs.recommended,
  ...tseslint.configs.recommendedTypeChecked,
  importPlugin.flatConfigs.recommended,
  {
    languageOptions: {
      parserOptions: {
        projectService: true,
        tsconfigRootDir: import.meta.dirname,
      },
    },
  },
  {
    files: ['**/*.js', '**/*.mjs'],
    ...tseslint.configs.disableTypeChecked,
  },
  {
    languageOptions: {
      globals: {
        ...globals.node,
        ...globals.jest,
      },
    },
    settings: {
      'import/parsers': {
        '@typescript-eslint/parser': ['.ts', '.tsx'],
        '@babel/eslint-parser': ['.js', '.mjs'],
      },
      'import/resolver': {
        typescript: {
          alwaysTryTypes: true,
        },
      },
      // Gives false positives because RRD has two versions in our dependency tree.
      'import/ignore': ['react-router-dom'],

      'import/internal-regex': aliasesRegex,
    },
    rules: {
      'import/order': [
        'error',
        {
          'newlines-between': 'always',
          groups: ['type', 'external', 'internal', ['parent', 'sibling', 'index'], 'object'],

          alphabetize: {
            order: 'asc',
          },
        },
      ],

      'import/no-cycle': 'error',

      'import/no-unresolved': [
        'error',
        {
          caseSensitiveStrict: true,
        },
      ],

      'workspaces/no-cross-imports': [
        'error',
        {
          allow: '@blink/core',
        },
      ],

      'no-restricted-imports': [
        'error',
        {
          patterns: [
            {
              group: ['*.css', '!*.module.css'],
              importNames: ['default'],
              message: 'For CSS modules, use the `.module.css` extension.',
            },
          ],
        },
      ],

    },
  },
}
```

Happy to try and create a proper repro repo. Do you have a base I can clone?

**Comment by carlocorradini** [comment]
The aliases are resolved only if the file is included from the corresponding tsconfig.json or jsconfig.json.
If you specificy an alias (i.e. A) in tsconfig.json, and file X has the alias but it's not present in `include` or `files` of tsconfig.json then it's not resolved.
How do you include the corresponding file in your configuration?

**Comment by RobHannay** [comment]
tsconfig just has a blanket

```
   "include": ["**/*.ts", "**/*.tsx"],
```

jsconfig has none of the above.

I can't work out what specifically in 3.8 has changed this behaviour

I've also tried adding 

```
  "include": ["**/*.js"],
```

to my `jsconfig` but it's not removed the error.

I know it's possible this could have been an unintentional feature, supporting alias resolution in JS files, but should it revert to Webpack resolution throughout ts/js files as that's what's really being done?

Is there a way that I can debug the resolved alias paths?

**Comment by carlocorradini** [comment]
Can you create a repro for a better context (Thanks)

**Comment by mkosir** [comment]
Not sure if the same issue 🤔 but I’m experiencing similar in project [react-parallax-tilt](https://github.com/mkosir/react-parallax-tilt). 
You can reproduce it by upgrading the [dependency](https://github.com/mkosir/react-parallax-tilt/blob/a90bb2bd8bfe3f72dd2075bd4df5c4e36a5ba745/package.json#L76) version to latest `3.8.3` and running `npm run lint`

**Comment by carlocorradini** [comment]
Mhmhm I see a lot of `../` in both `include` and `exclude`.
Tomorrow I'll try to find the cause.
@SuperchupuDev do you think we have something similar to the other issue?

**Comment by SuperchupuDev** [comment]
oh! @mkosir just to confirm, if you do `npm why tinyglobby` it shows up as 0.2.12, right?

@carlocorradini probably! it's definitely related to the same function :P but i need a minimal repro to investigate regardless

**Comment by estiller** [comment]
I'm using [NX](https://nx.dev/) as a monorepo tool, and I see the exact behaviour as well. NX relies on aliases in `tsconfig` to manage inter-package dependencies, and when I upgrade to `eslint-import-resolver-typescript` version `3.8.0` or above (including `3.8.3`) I get the exact same error as OP.

I ran `pnpm why tinyglobby`, and it indeed resolved to `0.2.12`. I tried overriding it to use `0.2.10`, but I got the same result.

**Comment by carlocorradini** [comment]
@estiller Can you provide a repro?
Thanks

**Comment by RobHannay** [comment]
Here's a repro https://github.com/RobHannay/typescript-import-resolver-repro

Version pinned to 3.7.0 which works fine. Bumping up to 3.8.3 causes the lint error in `app/app.js`

<img width="878" alt="Image" src="https://github.com/user-attachments/assets/a98f6ef1-b404-44b9-8d1d-9d3871a1cf21" />

**Comment by pralkarz** [comment]
> Not sure if the same issue 🤔 but I’m experiencing similar in project [react-parallax-tilt](https://github.com/mkosir/react-parallax-tilt). You can reproduce it by upgrading the [dependency](https://github.com/mkosir/react-parallax-tilt/blob/a90bb2bd8bfe3f72dd2075bd4df5c4e36a5ba745/package.json#L76) version to latest `3.8.3` and running `npm run lint`

I'm not sure about the other issues listed here, but I investigated this one in particular, and it seems like it's potentially a problem with `get-tsconfig`. After https://github.com/mkosir/react-parallax-tilt/blob/main/tsconfig.json is resolved by extending https://github.com/mkosir/react-parallax-tilt/blob/main/scripts/tsconfig.dev.json, `"include": ["../.", "../.storybook/**/*", "../.commitlintrc.ts"]` gets parsed as `[ '', '.storybook/**/*', '.commitlintrc.ts' ]` and passed to `tinyglobby` there: https://github.com/import-js/eslint-import-resolver-typescript/blob/b61fb4a894d20c38d1ac4656c395de2c30e4c27b/src/index.ts#L463.

The empty string doesn't match anything and the `no-unresolved` errors are raised. When manually patching that string to be `'.'` or `'./**/*'`, everything works fine.

A workaround could be to change `../.` to `.././**/*` (in which case `get-tsconfig` parses it as `**/*`), but a more robust fix should probably happen in `get-tsconfig`. I'm not sure why it parses `../.` into an empty string instead of `.`.

**Comment by carlocorradini** [comment]
@pralkarz Wow thanks for your discovery/help.
Tomorrow I'll try the other repro

**Comment by RobHannay** [comment]
> Here's a repro [RobHannay/typescript-import-resolver-repro](https://github.com/RobHannay/typescript-import-resolver-repro?rgh-link-date=2025-02-20T23%3A11%3A39.000Z)
> 
> Version pinned to 3.7.0 which works fine. Bumping up to 3.8.3 causes the lint error in `app/app.js`
> <img alt="Image" width="878" src="https://private-user-images.githubusercontent.com/609062/415410808-a98f6ef1-b404-44b9-8d1d-9d3871a1cf21.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDAwOTM4NDUsIm5iZiI6MTc0MDA5MzU0NSwicGF0aCI6Ii82MDkwNjIvNDE1NDEwODA4LWE5OGY2ZWYxLWI0MDQtNDRiOS04ZDFkLTlkMzg3MWExY2YyMS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwMjIwJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDIyMFQyMzE5MDVaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT04OTFhMDM0NWVmNTM5M2YxZTU2YzVmNGZjNTQxYzVkNjMzMDgwYTM5ZmMwM2IzYjBlMWExMGJjNzk4Y2IxODZlJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.sPe20aFH5f9iIRDqnTXjMy9jTCfaGgCmANlK0An0EZk">

Just went through a bunch of stuff to clean out some definitely-unrelated bits of the config and have pushed that up now too.

Here's the evidence of that repro:

<img width="814" alt="Image" src="https://github.com/user-attachments/assets/757b1a21-64f0-4017-a242-a7e89f89a6f3" />

**Comment by pralkarz** [comment]
> Here's a repro https://github.com/RobHannay/typescript-import-resolver-repro
> 
> Version pinned to 3.7.0 which works fine. Bumping up to 3.8.3 causes the lint error in `app/app.js`
> <img alt="Image" width="878" src="https://private-user-images.githubusercontent.com/609062/415410808-a98f6ef1-b404-44b9-8d1d-9d3871a1cf21.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDAwOTQyNTYsIm5iZiI6MTc0MDA5Mzk1NiwicGF0aCI6Ii82MDkwNjIvNDE1NDEwODA4LWE5OGY2ZWYxLWI0MDQtNDRiOS04ZDFkLTlkMzg3MWExY2YyMS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwMjIwJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDIyMFQyMzI1NTZaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1iODlkZTI3YTA5YTg4ZjMwYWVlZTBkYzU1ZWFhOWU2OWJlZTg2MjZjZmEzYWJkYjFmM2Y2OTg4MTg4YTZmNjc3JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.ERy_kCjW5RDqZ3aMZLaAuJsKlATpGFYlnoZhbezp59I">

I took a look at this one too. Seems like mapping aliases is handled differently in 3.8.x vs 3.7.0. In https://github.com/import-js/eslint-import-resolver-typescript/blob/b61fb4a894d20c38d1ac4656c395de2c30e4c27b/src/index.ts#L414, the mappers are created, but aren't used in any way while globbing, and only returned at the end. I'm not very familiar with this logic, so I can't gauge what goes wrong there, but it seems like it's unrelated to `tinyglobby`. 🤔 

CC: @carlocorradini @RobHannay @SuperchupuDev

**Comment by carlocorradini** [comment]
@pralkarz @RobHannay 
This is the correct behavior.
If `app.js` (or any other file) is not included in your `tsconfig.json`, how does that file determine the correct path mapping?
Version prior to `3.8` check if that path is known even if it is not found in any of the known `tsconfig.json` (or `jsconfig.json`) files. This raises multiple issues. One example is where the same mapping alias appears in various `tsconfig.json` files, but the destination (file referred to) differs. The latter is widely used in monorepo, where each package might have the same mapping alias (i.e. `@`), but the destination corresponds to the `src` directory of that package.
See issue #246

**Comment by RobHannay** [comment]
Thanks @carlocorradini, that makes sense! As I suspected then, in that we were actually relying on a bug.

I know bug fixes aren't usually considered "breaking" in the traditional sense, but I wonder if this (correct) behavioural change might cause breakages for a number of people that were relying on it.

Are you saying that adding `include` to me jsconfig should also match the `paths` in there to resolve this issue for me?

**Comment by JounQin** [comment]
@RobHannay Does VSCode/tsc itself complains no module resolved with your jsconfig/tsconfig settings? Yes so I think that's a bug fix for unintended behavior, but if not, we may need find out how does tsc fallback aliasing?

cc @carlocorradini

**Comment by RobHannay** [comment]
Both VS Code and WebStorm happily resolve the import paths.

https://github.com/user-attachments/assets/a01a58f7-ba7c-4ba1-8c95-aa614bb21722

**Comment by RobHannay** [comment]
Another small thing to note that I've just been looking at.

> How do you include the corresponding file in your configuration?

As I metioned, in our `jsconfig` we actually only have an `exclude` property, without `include`. According to the docs, that is sufficient:

https://code.visualstudio.com/docs/languages/jsconfig#_using-the-exclude-property

Adding `include` doesn't fix it anyway at the moment, and if that's what's needed then I'd be happy to add it. But according to documentation it shouldn't be necessary

**Comment by carlocorradini** [comment]
> Another small thing to note that I've just been looking at.
> 
> > How do you include the corresponding file in your configuration?
> 
> As I mentioned, in our `jsconfig` we actually only have an `exclude` property, without `include`. According to the docs, that is sufficient:
> 
> https://code.visualstudio.com/docs/languages/jsconfig#_using-the-exclude-property
> 
> Adding `include` doesn't fix it anyway at the moment, and if that's what's needed then I'd be happy to add it. But according to documentation it shouldn't be necessary

If no include and no files property are present we include everything ("**/*"). Same behavior as Typescript. Note that we apply your exclude if present

**Comment by carlocorradini** [comment]
> Both VS Code and WebStorm happily resolve the import paths.
> 
> https://github.com/user-attachments/assets/a01a58f7-ba7c-4ba1-8c95-aa614bb21722

The editor may be smart but if you compile with TSC the JS file Is not present in the outdir (can you try?)

**Comment by RobHannay** [comment]
> > Both VS Code and WebStorm happily resolve the import paths.
> > 
> >   
> >     
> >     
> > 
> >     Screen.Recording.2025-02-21.at.10.33.02.mov
> 
> The editor may be smart but if you compile with TSC the JS file Is not present in the outdir (can you try?)

It does get emitted by tsc

![Image](https://github.com/user-attachments/assets/523ed4de-0ab3-4d5a-8a29-f5a4432d3523)

**Comment by carlocorradini** [comment]
@RobHannay The .js file is emitted in the dist folder even though it's not present in both include and files? How is it possible?!
Can anyone explain this? I'm confused 😕

**Comment by RobHannay** [comment]
My guess would be that it's imported by a `.ts(x)` file at some point, so tsc needs to include it in the dist

**Comment by ehoogeveen-medweb** [comment]
Might be worth running tsc with [`--explainFiles`](https://www.typescriptlang.org/tsconfig/#explainFiles) (or setting `"explainFiles": true` in the compiler options) to see what it says.

**Comment by RobHannay** [comment]
Nice. Well I can see it's just imports as suspected:

```
(blah/blah)/components/Primer/Primer.js
  Imported via './components/Primer/Primer' from file 'webapp/src/components/Login/Login.js'
```

**Comment by JounQin** [comment]
> So when no matched `project` found it still should fallback to its nearest listed `project`, how do you think? [@carlocorradini](https://github.com/carlocorradini?rgh-link-date=2025-02-21T10%3A49%3A10.000Z)

@carlocorradini Fallback to nearest `tsconfig.json` is desired behavior IMO.

**Comment by carlocorradini** [comment]
We have two different types of issues:
1. Thanks to @pralkarz, we discovered that if `include` contains `../.` (or `.`), `get-tsconfig` returns an empty string (`""`).
   I've opened an issue: <https://github.com/privatenumber/get-tsconfig/issues/88>
   @mkosir The change from `../.` to `.././**/*` works. You can modify it or wait till the `get-tsconfig` issue is fixed.
1. @RobHannay To fix this, we need to modify from line https://github.com/import-js/eslint-import-resolver-typescript/blob/b61fb4a894d20c38d1ac4656c395de2c30e4c27b/src/index.ts#L309
   `Pseudocode`:
   _If_ `mappedPath` (resolved file path) already exists in the associated `Set`: do nothing.
   _Otherwise_, we add this new file (`mappedPath`) to the associated `Set` to ensure that future resolutions have the file matched with the corresponding (correct) mapping function.

What are your thoughts? 🫡

I'll try to fix this in a few days 😋

**Comment by carlocorradini** [comment]
@JounQin See <https://github.com/carlocorradini/eslint-import-resolver-typescript/tree/363>
`mapping` is a `Map` where the key is the absolute (native) path of a file and the value is a `Set` of paths matching functions (applicable).
When the `getMappedPath` function is called, we ensure that each `mappedPath` exists in the mapping: If not, add the new path with the corresponding functions; otherwise, add any (possible) missing functions.
Everything works however there is catch:
Each file that has to be linted is (mostly) evaluated at random (try this with `yarn run test:importTransitive`). As a result, a file that is transitively imported from another and does not yet exist in the mapping may be evaluated first, returning not found. However, subsequently, following the import tree (in this example from `index.ts`), we properly populate mapping with the correct file paths/functions, and all mapping are fully resolved: You can see this in the log.
This is because import/import-x calls resolve once, therefore correct path resolution (later) from the same file is discarded/forgotten.
To solve this, if the file is not in the mapping, we attempt to resolve the path mapping using all known functions. However, this can produce numerous different possible paths for the same alias, causing the same as issue #246. Or we can force to list all files in the `tsconfig.json`.
Which one do you prefer?

**Comment by JounQin** [comment]
> To solve this, if the file is not in the mapping, we attempt to resolve the path mapping using all known functions. However, this can produce numerous different possible paths for the same alias, causing the same as issue #246. Or we can force to list all files in the tsconfig.json.
Which one do you pre#246 

@carlocorradini If the file is not in the mapping, can we pretend the nearest `tsconfig.json` should match?

**Comment by carlocorradini** [comment]
@JounQin I am not sure whether the nearest `tsconfig.json` is a viable answer... we can try every possible mapping(s) (like pre-3.8) if the current file does not exist in any set.

**Comment by ehoogeveen-medweb** [comment]
FWIW, using the nearest `tsconfig.json` is [how `project: true` works](https://typescript-eslint.io/packages/parser#project) in typescript-eslint. I don't know whether it keeps going up the chain if the nearest `tsconfig.json` doesn't include the file - the documentation doesn't say.

**Comment by JounQin** [comment]
> [@JounQin](https://github.com/JounQin?rgh-link-date=2025-03-10T09%3A26%3A38.000Z) I am not sure whether the nearest `tsconfig.json` is a viable answer... we can try every possible mapping(s) (like pre-3.8) if the current file does not exist in any set.

Peronally I think nearest resolving is fine enough, we can try and see feedbacks?

**Comment by JounQin** [comment]
@carlocorradini Are you working this one?

**Comment by carlocorradini** [comment]
> [@carlocorradini](https://github.com/carlocorradini) Are you going to work on this one?

👍

**Comment by carlocorradini** [comment]
See PR #372

**Comment by donaldpipowitch** [comment]
Mhh, I have `".storybook/**/*",` in my include and all file imports to that directory return a `Unable to resolve path to module`. It is the only case that I have that starts with a `.`. Other imports work fine and if I rename it, it also works fine. But this is Storybooks default directory. Is this a regression of this issue?

**Comment by JounQin** [comment]
Please open a new issue with minimal but runnable reproduction instead.
